### DISCUSSION TITLE
Runtime for ensemble-stat

### DISCUSSION THREAD

**Discussion by michelleharrold** [dtcenter/METplus#1606] [File I/O] [answered]
Hello!

We are running ensemble-stat (METv10.1.0) on a 36-h forecast with hourly output for select fields (e.g., accum. precip, reflectivity, and temp). For a 36-h forecast, we are experiencing wall clock times around 6 hours per variable, which seems long (longer than the model to run!). Is this anticipated? Is there anything we can do to decrease these times?

Thanks!
Michelle Harrold

**Comment by j-opatz** [comment]
Hi Michelle,

And thank you for your inquiry. While you are processing a lot of data (4 fields, 36 ensembles, higher resolution), your METplus run time seems to be higher than expected/acceptable and is unexpected to hear.

Do you have a sense of how long a similar process would take to run on older MET versions (pre-MET v10.1.0), for comparison? 

If you can point me to some sample data, I could take a look and see if a pared-down setup shows time increases between version levels/new logic. As you said, waiting for the verification longer than the model output is something we definitely want to correct!

**Comment by michelleharrold** [comment]
Hi John,

Thanks a bunch for offering to take a look at some of my config files; I really appreciate it.

I wanted to at least respond and let you know I am getting back to this today -- I am currently running a timing test on hera using METv10.0.0 and METv10.1.0 to see if there is an increase in run-time between the two versions. I will report back once I have some more info.

Thanks!
Michelle

**Comment by michelleharrold** [comment]
Hi John,

I do not see a difference in run time between METv10.0.0 and METv10.1.0 when running ensemble-stat; however, I do see a very large difference when running with 25-km model output vs. 3-km model output, so my guess is this is merely an issue with having more data to process.

If you would like to take a look at our METplus configure file, it is available via [GitHub](https://github.com/michelleharrold/regional_workflow/blob/feature/breakout_io/ush/templates/parm/metplus/EnsembleStat_APCP01h.conf) (linking to 1-h accumulated precip as an example). Each forecast hour takes about 10 minutes to process in ensemble-stat, so a 36-h hour forecast takes ~6 hours of wall clock time. We currently have one job submission for all 36 hours; we have discussed submitting each forecast hour as a separate job, but in the past job schedulers have limited how many active jobs we can have (and with all of our MET vx, there would be A LOT of jobs!)...so being able to run the 36-h forecast in one submission would be optimal.

Any pro tips would be greatly welcomed!
Michelle

**Comment by j-opatz** [comment]
Hi Michelle, 

Thanks for the followup on this, including the assessment for 25km to 3km time changes. It's disappointing to hear that the tool isn't where we'd like it to be re: speed, so we're still looking into this to see if any optimization can be accomplished.

We discussed this at our last MET team meeting and @jprestop has offered to take a look at the installation to see if there are any bug flags that might be slowing things down.

Would you be able to share where this installation is/where the work is being done? Thanks!

**Comment by michelleharrold** [comment]
Hi John and Julie,

Really appreciate all the help with this!

I am running on Hera (NOAA) using this installation of METv10.1.0 (and associated module files, etc):
/contrib/met/10.1.0

Thanks!
Michelle

**Reply by jprestop** [reply]
Thanks, @michelleharrold.

I noticed in MET's config.log file that CFLAGS was set to "-g -O2" with met-10.1.0 even though I wasn't setting CFLAGS at all.  I installed 10.1.2 and et CFLAGS to "".  That seems to have worked as I don't see any instances of "-g -O2" in MET's config.log file.  Could you please try running the following and testing with 10.1.2 and see if this problem still exists?
```
module use -a /contrib/anaconda/modulefiles
module load intel/2022.1.2
module load anaconda/latest
module use -a /contrib/met/modulefiles/
module load met/10.1.2
```

**Comment by michelleharrold** [comment]
@jprestop - thank you for recompiling METv10.1.2 on hera!

I just ran a test, and I saw a very minimal impact -- it did improve by a few seconds,  but ensemble-stat still takes ~10 minutes per forecast hour.

**Reply by jprestop** [reply]
Thank you, @michelleharrold.  I'm sorry to hear that it only improved by a few seconds.  That is discouraging.  @j-opatz, it looks like the compilation to remove "-g -O2" did not work to speed up the ensemble-stat run.

**Comment by j-opatz** [comment, marked as answer]
@michelleharrold,

I wanted to touch base with you on this as it's been a while since any progress has been made. 

I discussed this with the METplus team at large and unfortunately, there doesn't seem to be a "light bulb" solution to this. We still agree that the run time you're experiencing is inadequate. But this will require a partial or complete tool optimization which calls for time and funding dedication. Both of which, as I'm sure you know, are in short supply.

This is not to say that this won't be investigated or pursued, but given the nature of the work needed (e.g. general improvement of tools), we can't put this in any definite beta.

I'm going to write up an issue for this and link it here so you have something to track this.

**Reply by j-opatz** [reply]
@michelleharrold 

Apologies that it took so long, but I've constructed an [issue](https://github.com/dtcenter/MET/issues/2184) and placed it in the "backlog of development ideas" milestone. It also mentions this Discussion thread so you can find your way back to it when we start to work on these items.

I sincerely hope that we can find a solution to this and not prolong the 6+ hour method you're currently utilizing.
